model/PermissionsService.php

```diff
@@ -54,55 +56,150 @@ public function __construct(
         $this->eventManager = $eventManager;
     }
 
-    public function saveResourcePermissionsRecursive(
+    /**
+     * Returns an opaque array holding the list of actions to be performed
+     * (i.e. a permissions diff) to fulfill a user's change permissions request.
+     */
+    public function getNormalizedRequest(
         core_kernel_classes_Resource $resource,
         array $privilegesToSet
+    ): array {
+        return $this->strategy->normalizeRequest(
+            $this->dataBaseAccess->getResourcePermissions($resource->getUri()),
+            $privilegesToSet
+        );
+    }
+
+    // @todo Check unit tests
+
+    public function savePermissions(
+        bool $isRecursive,
+        core_kernel_classes_Class $class,
+        array $privilegesToSet
     ): void {
-        $this->saveResourcePermissions($resource, $privilegesToSet, true);
+        $this->setPermissionsByResourceRootAndTriggerEvents(
```

Comment:
im curious, if theres a method called setPermissionsByResourceRootAndTriggerEvents this means theres a setPermissionsBySomethingElse somewhere else? :)